plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -0,0 +1,614 @@
+import org.sonatype.nexus.repository.FacetSupport;
+import org.sonatype.nexus.repository.content.InvalidContentException;
+import org.sonatype.nexus.repository.maven.internal.Path.HashType;
+import org.sonatype.nexus.repository.maven.internal.policy.ChecksumPolicy;
+import org.sonatype.nexus.repository.maven.internal.policy.VersionPolicy;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.repository.storage.StorageTx;
+import org.sonatype.nexus.repository.util.NestedAttributesMap;
+import org.sonatype.nexus.repository.view.payloads.BlobPayload;
+import org.sonatype.nexus.repository.view.payloads.StringPayload;
+
+import com.google.common.base.Strings;
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.hash.HashCode;
+import com.tinkerpop.blueprints.Direction;
+import com.tinkerpop.blueprints.Vertex;
+import com.tinkerpop.blueprints.impls.orient.OrientVertex;
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * A {@link MavenFacet} that persists Maven artifacts and metadata to a {@link StorageFacet}.
+ * <p/>
+ * Structure for artifacts (CMA components and assets):
+ * <ul>
+ * <li>CMA components: keyed by groupId:artifactId:version</li>
+ * <li>CMA assets: keyed by path</li>
+ * </ul>
+ * <p/>
+ * Structure for metadata (CMA assets only):
+ * <ul>
+ * <li>CMA assets: keyed by path</li>
+ * </ul>
+ * In both cases, "external" hashes are stored as separate asset, as their path differs too.
+ *
+ * @since 3.0
+ */
+@Named
+public class MavenFacetImpl
+    extends FacetSupport
+    implements MavenFacet
+{
+  // artifact shared properties of both, artifact component and artifact asset
+
+  private static final String P_GROUP_ID = "groupId";
+
+  private static final String P_ARTIFACT_ID = "artifactId";
+
+  private static final String P_VERSION = "version";
+
+  private static final String P_BASE_VERSION = "baseVersion";
+
+  private static final String P_CLASSIFIER = "classifier";
+
+  private static final String P_EXTENSION = "extension";
+
+  // artifact component properties
+
+  private static final String P_COMPONENT_KEY = "key"; // component key: G:A:V for fast look-ups
+
+  // shared properties for both artifact and metadata assets
+
+  private static final String P_ASSET_KEY = "key"; // asset key: G:A:V[:C]:E for fast look-ups
+
+  private static final String P_LAST_MODIFIED = "lastModified";
+
+  private static final String P_ASSET_KIND = "kind"; // denoting asset kind
+
+  private static final String ASSET_KIND_ARTIFACT = "artifact";
+
+  private static final String ASSET_KIND_METADATA = "metadata";
+
+  // ==
+
+  private final MimeSupport mimeSupport;
+
+  public static final String CONFIG_KEY = "maven";
+
+  // ==
```

Comment:
here and above pointless "// ==" markers, arent' clearly providing any organization value, only noise